File: scripts/solar/solar-common/util/game/PlatRandomNumUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 1,["7"] = 1,["8"] = 2,["9"] = 2,["10"] = 11,["11"] = 11,["12"] = 11,["14"] = 11,["15"] = 15,["16"] = 16,["17"] = 17,["18"] = 18,["19"] = 15,["20"] = 21,["21"] = 22,["22"] = 23,["24"] = 25,["25"] = 26,["26"] = 27,["27"] = 28,["28"] = 29,["31"] = 32,["32"] = 33,["33"] = 34,["34"] = 37,["35"] = 38,["37"] = 40,["39"] = 42,["40"] = 44,["41"] = 45,["42"] = 46,["44"] = 48,["45"] = 49,["46"] = 21,["47"] = 12});
local ____exports = {}
local ____PlatUtil = require("solar.solar-common.util.game.PlatUtil")
local PlatUtil = ____PlatUtil.default
local ____ArchiveUtil = require("solar.solar-common.util.archive.ArchiveUtil")
local ArchiveUtil = ____ArchiveUtil.default
____exports.default = __TS__Class()
local PlatRandomNumUtil = ____exports.default
PlatRandomNumUtil.name = "PlatRandomNumUtil"
function PlatRandomNumUtil.prototype.____constructor(self)
end
function PlatRandomNumUtil.getRandomNumber(self, whichPlayer, key)
    local nowIndex = ArchiveUtil:get(whichPlayer, "i_" .. key) or 0
    local realKey = (tostring(nowIndex) .. "X") .. key
    return PlatUtil:getBackendLogicIntResult(whichPlayer, realKey)
end
function PlatRandomNumUtil.updateRandomNumber(self, whichPlayer, key, groupkey)
    if not IsHandle(whichPlayer) then
        return false
    end
    local limitKey = (tostring(GetPlayerId(whichPlayer)) .. "_") .. key
    if ____exports.default.limitKey[limitKey] == true then
        log.errorWithTraceBack("一局游戏中不能更新2次随机数!")
        if not isDebug then
            return false
        end
    end
    ____exports.default.limitKey[limitKey] = true
    local nowIndex = ArchiveUtil:get(whichPlayer, "i_" .. key) or 0
    local realKey = (tostring(nowIndex) .. "X") .. key
    if nowIndex == 0 or nowIndex == "0" or nowIndex == "" then
        nowIndex = 1
    else
        nowIndex = 0
    end
    ArchiveUtil:set(whichPlayer, "i_" .. key, nowIndex)
    if not PlatUtil:checkBackendLogicExists(whichPlayer, realKey) then
        PlatUtil:requestBackendLogic(whichPlayer, "0X" .. key, groupkey)
        return PlatUtil:requestBackendLogic(whichPlayer, "1X" .. key, groupkey)
    end
    PlatUtil:removeBackendLogicResult(whichPlayer, realKey)
    return PlatUtil:requestBackendLogic(whichPlayer, realKey, groupkey)
end
PlatRandomNumUtil.limitKey = {}
return ____exports

File: scripts/solar/solar-common/util/game/PlatRandomNumUtil.js
import PlatUtil from "@/PlatUtil";
import ArchiveUtil from "@/ArchiveUtil";
/**
 * (如果平台的获得最新的随机数事件有延迟0.1s的话 可以考虑采用2组随机数来模拟立即获得最新随机数 基本没什么用)
 * 采用2组随机数来获得当前生成的值。相当于每局只能更新一次随机数 并立即获得结果 (实际就是切换使用的另一组)
 * 平台随机数工具
 * key 不支持下划线 只能大小写字母和数字组合
 * @deprecated see PlatUtil.getBackendLogicIntResult 推荐使用原始的随机数
 */
export default class PlatRandomNumUtil {
    static limitKey = {};
    static getRandomNumber(whichPlayer, key) {
        let nowIndex = ArchiveUtil.get(whichPlayer, "i_" + key) || 0;
        let realKey = nowIndex + "X" + key;
        return PlatUtil.getBackendLogicIntResult(whichPlayer, realKey);
    }
    static updateRandomNumber(whichPlayer, key, groupkey) {
        if (!IsHandle(whichPlayer)) {
            return false;
        }
        let limitKey = GetPlayerId(whichPlayer) + "_" + key;
        if (PlatRandomNumUtil.limitKey[limitKey] == true) {
            log.errorWithTraceBack("一局游戏中不能更新2次随机数!");
            if (!isDebug) {
                return false;
            }
        }
        PlatRandomNumUtil.limitKey[limitKey] = true;
        let nowIndex = ArchiveUtil.get(whichPlayer, "i_" + key) || 0;
        let realKey = nowIndex + "X" + key;
        //
        if (nowIndex == 0 || nowIndex == "0" || nowIndex == "") {
            nowIndex = 1;
        }
        else {
            nowIndex = 0;
        }
        ArchiveUtil.set(whichPlayer, "i_" + key, nowIndex);
        //说明是第一次生成 则2个都生成
        if (!PlatUtil.checkBackendLogicExists(whichPlayer, realKey)) {
            PlatUtil.requestBackendLogic(whichPlayer, "0X" + key, groupkey);
            return PlatUtil.requestBackendLogic(whichPlayer, "1X" + key, groupkey);
        }
        PlatUtil.removeBackendLogicResult(whichPlayer, realKey);
        return PlatUtil.requestBackendLogic(whichPlayer, realKey, groupkey);
    }
}
